### ISSUE TITLE
Timescale issues

### ISSUE THREAD

**Issue by pg9182** [pg9182/northstar-dedicated#89]
I'm just going to dump some stuff about the timescale issues with the nodrv patch and/or new wine versions here.

**Comment by pg9182** [comment]
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/3cfe10d6-4606-481a-b259-741ffddbe1c4)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/ebea9d09-bdfc-4e17-ab30-66fec0e13ca1)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/36911402-afae-4006-87b5-e98079234f3d)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/07efe62d-e8a8-4440-9e88-93ecfc684f15)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/57092486-20c3-4c8f-8821-9765abea4eab)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/dcaaf74f-5738-42e3-b5db-151ce51bc2d1)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/bd6747c8-c2ce-4aa6-b5fa-e560e3294bd2)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/f7cc206f-cd1d-4f32-8fde-8948c85d135b)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/372c6ccc-907b-474e-9578-95f3ddf8d391)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/53ac0136-6a12-40a3-8007-1ec18b755f5f)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/b8516330-fb61-451d-a65f-b791edec31e2)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/cf70e69b-47a7-49d2-b6ed-cd92bc8300c6)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/a3750551-c6f2-45cd-853f-c9401b26cda3)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/4a6eaef5-a9f4-43d7-b1d6-54e1ac9a1701)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/4edbac26-b8aa-4b31-a901-ce472e543a2f)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/d4051a0b-f571-4e88-8007-aba4fed5454a)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/e54590ca-374a-4f7b-9217-895f1cf6507b)
![image](https://github.com/pg9182/northstar-dedicated/assets/96569817/e180c392-815d-475f-8d61-064e32a376b9)

**Comment by pg9182** [comment]
TLDR: we can avoid this entire issue while also making everything else much easier by using wine-9.0 with headless weston.

```
export WAYLAND_DISPLAY=wayland-1 XDG_RUNTIME_DIR=/tmp XDG_SESSION_TYPE=wayland
weston --backend=headless --use-pixman --no-config &

wine64 wineboot --init

wine64 reg add 'HKCU\Software\Wine' /v 'Version' /t REG_SZ /d 'win10' /f
wine64 reg add 'HKCU\Software\Wine\Drivers' /v 'Audio' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\WineDbg' /v 'ShowCrashDialog' /t REG_DWORD /d 0 /f
wine64 reg add 'HKCU\Software\Wine\Drivers' /v 'Graphics' /t REG_SZ /d 'wayland' /f

wine64 reg add 'HKLM\System\CurrentControlSet\Services\WineBus' /v 'DisableHidraw' /t REG_DWORD /d 1 /f
wine64 reg add 'HKLM\System\CurrentControlSet\Services\WineBus' /v 'DisableInput' /t REG_DWORD /d 1 /f
wine64 reg add 'HKLM\System\CurrentControlSet\Services\WineBus' /v 'Enable SDL' /t REG_DWORD /d 0 /f

wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'mscoree' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'mshtml' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'winemenubuilder' /t REG_SZ /d '' /f

wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'd3d9' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'd3d10' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'd3d11' /t REG_SZ /d 'native' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'd3d12' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'wined3d' /t REG_SZ /d '' /f
wine64 reg add 'HKCU\Software\Wine\DllOverrides' /v 'winevulkan' /t REG_SZ /d '' /f

nswrap /path/to/northstar
```

**Comment by pg9182** [comment]
I'm going to work on the new docker image now, which will feature:

- wine 9
- weston/wayland (instead of xvfb)
- debian base image (instead of alpine) (for better host compatibility)
- ability to download updated northstar versions / mods / ci builds (replacing northstar-dedicated-ci)
- better layering logic (probably from scratch instead of trying to coerce buildkit into doing what we want)
- dropping legacy northstar support
- possibly thunderstore integration
- support executing concommands via stdin (via WM_COPYDATA)
- better support for graceful client disconnect on shutdown (via WM_QUIT and/or the concommand)
- better integration with the main northstar release pipeline

Will probably also make some launcher changes for the next release:

- automatically exit on registration failure (to make it easier to manage the container)
- send container metadata during server registration (for troubleshooting and stats)

**Comment by pg9182** [comment]
Also going to consider making the image support being run as a systemd portable service, and being bootable as an ISO with config via cloud-init.

**Comment by pg9182** [comment]
Will probably just do a custom wlroots-based compositor from scratch since it'll only be a few hundred lines, it'll have fewer dependencies than weston, and it'll be nice to have it integrated with the rest of nswrap.

**Comment by pg9182** [comment]
This is all that's necessary for creating a dummy compositor for wayland with wine 9.0-rc2.

```c
// pkg-config --cflags --libs pixman-1 wlroots wayland-server
// wayland-scanner server-header /usr/share/wayland-protocols/stable/xdg-shell/xdg-shell.xml src/nswrap/xdg-shell-protocol.h
// wayland-scanner server-header /usr/share/wayland-protocols/unstable/pointer-constraints/pointer-constraints-unstable-v1.xml src/nswrap/pointer-constraints-unstable-v1-protocol.h
// gcc -Wall -Wextra -I/usr/include/pixman-1 -I/usr/include/libdrm src/nswrap/nswrap2.c -lpixman-1 -lwlroots -lwayland-server

#define WLR_USE_UNSTABLE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <wayland-server-core.h>
#include <wlr/backend.h>
#include <wlr/backend/session.h>
#include <wlr/backend/headless.h>
#include <wlr/render/wlr_renderer.h>
#include <wlr/render/pixman.h>
#include <wlr/types/wlr_compositor.h>
#include <wlr/types/wlr_xdg_shell.h>
#include <wlr/types/wlr_subcompositor.h>
#include <wlr/types/wlr_pointer_constraints_v1.h>
#include <wlr/types/wlr_relative_pointer_v1.h>
#include <wlr/util/log.h>

extern struct wlr_allocator *wlr_shm_allocator_create(void);

int main(void) {
    wlr_log_init(WLR_DEBUG, NULL);
    struct wl_display *display = wl_display_create();
    struct wlr_backend *backend = wlr_headless_backend_create(display);
    if (!backend) {
        wlr_log(WLR_ERROR, "Failed to create wlr_headless_backend");
        exit(1);
    }

    struct wlr_renderer *renderer = wlr_pixman_renderer_create();
    if (!renderer) {
        wlr_log(WLR_ERROR, "Failed to create wlr_pixman_renderer");
        exit(1);
    }

    struct wlr_compositor *compositor = wlr_compositor_create(display, renderer);
    if (!compositor) {
        wlr_log(WLR_ERROR, "Failed to create wlr_compositor");
        exit(1);
    }

    // required for winewayland.drv
    if (!wlr_renderer_init_wl_shm(renderer, display)) {
        wlr_log(WLR_ERROR, "Failed to initialize wl_shm on renderer");
        exit(1);
    }

    // required for winewayland.drv
    struct wlr_subcompositor *subcompositor = wlr_subcompositor_create(display);
    if (!subcompositor) {
        wlr_log(WLR_ERROR, "Failed to create wlr_subcompositor");
        exit(1);
    }

    // required for winewayland.drv
    struct wlr_pointer_constraints_v1 *pointer_constraints_v1 = wlr_pointer_constraints_v1_create(display);
    if (!pointer_constraints_v1) {
        wlr_log(WLR_ERROR, "Failed to create wlr_pointer_constraints_v1");
        exit(1);
    }

    // required for winewayland.drv
    struct wlr_relative_pointer_manager_v1 *relative_pointer_manager_v1 = wlr_relative_pointer_manager_v1_create(display);
    if (!relative_pointer_manager_v1) {
        wlr_log(WLR_ERROR, "Failed to create wlr_relative_pointer_manager_v1");
        exit(1);
    }

    // required for winewayland.drv
    struct wlr_xdg_shell *xdg_shell = wlr_xdg_shell_create(display, 5);
    if (!xdg_shell) {
        wlr_log(WLR_ERROR, "Failed to create wlr_xdg_shell");
        exit(1);
    }

    if (wl_display_add_socket(display, "/tmp/nswrap-wayland")) {
        wlr_log(WLR_ERROR, "Failed to create socket");
        exit(1);
    }

    if (!wlr_backend_start(backend)) {
        wlr_log(WLR_ERROR, "Failed to start backend");
        wlr_backend_destroy(backend);
        exit(1);
    }

    wlr_log(WLR_INFO, "WAYLAND_DISPLAY=/tmp/nswrap-wayland");
    wl_display_run(display);
    wl_display_destroy(display);
    wlr_renderer_destroy(renderer);
}
```

**Comment by pg9182** [comment]
https://github.com/pg9182/NorthstarDocker
